File: compatibility-tests/sdk-test-go/tests/gcs_test.go
package tests

import (
	"bytes"
	"context"
	"io"
	"testing"

	"floci-gcp-sdk-test-go/internal/testutil"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestGCS(t *testing.T) {
	ctx := context.Background()
	client := testutil.StorageClient(ctx)
	defer client.Close()

	projectID := testutil.ProjectID()
	bucketName := uniqueName("go-gcs")
	objectName := "test-object.txt"
	content := "Hello, GCP Cloud Storage from Go!"

	t.Cleanup(func() {
		bkt := client.Bucket(bucketName)
		bkt.Object(objectName).Delete(ctx)
		bkt.Object("copy-" + objectName).Delete(ctx)
		bkt.Delete(ctx)
	})

	t.Run("CreateBucket", func(t *testing.T) {
		err := client.Bucket(bucketName).Create(ctx, projectID, nil)
		require.NoError(t, err)
	})

	t.Run("ListBuckets", func(t *testing.T) {
		it := client.Buckets(ctx, projectID)
		found := false
		for {
			attrs, err := it.Next()
			if err != nil {
				break
			}
			if attrs.Name == bucketName {
				found = true
				break
			}
		}
		assert.True(t, found, "created bucket should appear in list")
	})

	t.Run("UploadObject", func(t *testing.T) {
		w := client.Bucket(bucketName).Object(objectName).NewWriter(ctx)
		w.ContentType = "text/plain"
		_, err := io.WriteString(w, content)
		require.NoError(t, err)
		require.NoError(t, w.Close())
	})

	t.Run("DownloadObject", func(t *testing.T) {
		r, err := client.Bucket(bucketName).Object(objectName).NewReader(ctx)
		require.NoError(t, err)
		defer r.Close()

		var buf bytes.Buffer
		_, err = buf.ReadFrom(r)
		require.NoError(t, err)
		assert.Equal(t, content, buf.String())
	})

	t.Run("ObjectMetadata", func(t *testing.T) {
		attrs, err := client.Bucket(bucketName).Object(objectName).Attrs(ctx)
		require.NoError(t, err)
		assert.Equal(t, objectName, attrs.Name)
		assert.Equal(t, "text/plain", attrs.ContentType)
		assert.Equal(t, int64(len(content)), attrs.Size)
	})

	t.Run("ListObjects", func(t *testing.T) {
		it := client.Bucket(bucketName).Objects(ctx, nil)
		found := false
		for {
			attrs, err := it.Next()
			if err != nil {
				break
			}
			if attrs.Name == objectName {
				found = true
				break
			}
		}
		assert.True(t, found, "uploaded object should appear in list")
	})

	t.Run("CopyObject", func(t *testing.T) {
		src := client.Bucket(bucketName).Object(objectName)
		dst := client.Bucket(bucketName).Object("copy-" + objectName)
		_, err := dst.CopierFrom(src).Run(ctx)
		require.NoError(t, err)

		r, err := dst.NewReader(ctx)
		require.NoError(t, err)
		defer r.Close()

		var buf bytes.Buffer
		_, err = buf.ReadFrom(r)
		require.NoError(t, err)
		assert.Equal(t, content, buf.String())
	})

	t.Run("DeleteObject", func(t *testing.T) {
		err := client.Bucket(bucketName).Object(objectName).Delete(ctx)
		require.NoError(t, err)
	})

	t.Run("DeleteBucket", func(t *testing.T) {
		client.Bucket(bucketName).Object("copy-" + objectName).Delete(ctx)
		err := client.Bucket(bucketName).Delete(ctx)
		require.NoError(t, err)
	})
}

File: compatibility-tests/sdk-test-node/tests/gcs.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { Storage } from '@google-cloud/storage';
import { ENDPOINT, PROJECT_ID, uniqueName } from './setup';

describe('Cloud Storage (GCS)', () => {
  let storage: Storage;
  let bucketName: string;
  const objectName = 'test-object.txt';
  const objectContent = 'Hello, GCP Cloud Storage from Node.js!';

  beforeAll(() => {
    storage = new Storage({
      apiEndpoint: ENDPOINT,
      projectId: PROJECT_ID,
    });
    bucketName = uniqueName('test-bucket');
  });

  afterAll(async () => {
    try {
      await storage.bucket(bucketName).file(objectName).delete().catch(() => {});
      await storage.bucket(bucketName).delete().catch(() => {});
    } catch {
      // ignore cleanup errors
    }
  });

  it('should create a bucket', async () => {
    const [bucket] = await storage.createBucket(bucketName);
    expect(bucket.name).toBe(bucketName);
  });

  it('should list buckets and find created bucket', async () => {
    const [buckets] = await storage.getBuckets();
    expect(buckets.some((b) => b.name === bucketName)).toBe(true);
  });

  it('should upload an object', async () => {
    await storage.bucket(bucketName).file(objectName).save(objectContent, {
      contentType: 'text/plain',
    });
  });

  it('should download and verify object content', async () => {
    const [content] = await storage.bucket(bucketName).file(objectName).download();
    expect(content.toString()).toBe(objectContent);
  });

  it('should get object metadata', async () => {
    const [metadata] = await storage.bucket(bucketName).file(objectName).getMetadata();
    expect(metadata.name).toBe(objectName);
    expect(metadata.contentType).toBe('text/plain');
    expect(Number(metadata.size)).toBeGreaterThan(0);
  });

  it('should list objects in bucket', async () => {
    const [files] = await storage.bucket(bucketName).getFiles();
    expect(files.some((f) => f.name === objectName)).toBe(true);
  });

  it('should copy an object', async () => {
    const destName = 'test-object-copy.txt';
    await storage.bucket(bucketName).file(objectName).copy(
      storage.bucket(bucketName).file(destName)
    );
    const [exists] = await storage.bucket(bucketName).file(destName).exists();
    expect(exists).toBe(true);
    await storage.bucket(bucketName).file(destName).delete();
  });

  it('should delete object', async () => {
    await storage.bucket(bucketName).file(objectName).delete();
    const [exists] = await storage.bucket(bucketName).file(objectName).exists();
    expect(exists).toBe(false);
  });

  it('should delete bucket', async () => {
    await storage.bucket(bucketName).delete();
    const [exists] = await storage.bucket(bucketName).exists();
    expect(exists).toBe(false);
    bucketName = '';
  });
});
